File: events/pubsub.go
package events

import (
	"github.com/nats-io/nats.go"
	"github.com/rs/zerolog/log"
)

type PubSub struct {
	conn *nats.EncodedConn
}

func NewPubSub(addr string) PubSub {
	conn, err := nats.Connect(addr)

	if err != nil {
		log.Fatal().Err(err).Msg("PubSub.Init")
	}

	enc, err := nats.NewEncodedConn(conn, nats.JSON_ENCODER)

	if err != nil {
		log.Fatal().Err(err).Msg("PubSub.NATS.Encoder")
	}

	return PubSub{enc}
}

func (p PubSub) Close() {
	p.conn.Close()
}

func (p *PubSub) Subscribe(event string, handler interface{}) *nats.Subscription {
	sub, err := p.conn.Subscribe(event, handler)

	if err != nil {
		log.Error().Err(err).Str("event", event).Msg("PubSub.Subscribe")
	}

	return sub
}

func (p *PubSub) Publish(event string, payload interface{}) {
	err := p.conn.Publish(event, payload)

	if err != nil {
		log.Error().Err(err).Str("event", event).Msg("PubSub.Publish")
	}
}

File: services/web/src/utils/pubsub.ts
import {
  Codec,
  connect,
  Msg,
  NatsConnection,
  NatsError,
  StringCodec,
  Subscription,
} from 'nats.ws';

const SERVER_URL = 'ws://localhost:8080/nats';

const AUTH = {
  user: import.meta.env.VITE_NATS_USER,
  pass: import.meta.env.VITE_NATS_PASS,
};

export class PubSub {
  private codec: Codec<string>;
  private conn: NatsConnection | null = null;

  constructor() {
    this.codec = StringCodec();
  }

  init = async (): Promise<void> => {
    try {
      this.conn = await connect({
        servers: SERVER_URL,
        user: AUTH.user,
        pass: AUTH.pass,
      });
    } catch (err) {
      console.error(err);
    }
  };

  subscribe = <T>(
    event: string,
    cb: (data: T) => void
  ): Subscription | undefined => {
    const callback = (err: NatsError | null, msg: Msg): void => {
      if (err) {
        throw err;
      }
      const data: T = JSON.parse(this.codec.decode(msg.data));
      cb(data);
    };

    const subscription = this.conn?.subscribe(event, { callback });
    return subscription;
  };
}
